File: backend/src/main/java/com/example/smart/lighting/scenes/with_natural/language/entity/Schedule.java
package com.example.smart.lighting.scenes.with_natural.language.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.type.SqlTypes;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Entity representing a scheduled automation rule.
 *
 * <p>Schedules define time-based or event-based triggers for lighting actions.
 * They support various trigger types:</p>
 * <ul>
 *   <li><b>time</b> - Execute at specific times with weekday filtering</li>
 *   <li><b>sun</b> - Execute relative to sunrise/sunset</li>
 *   <li><b>sensor</b> - Execute based on sensor thresholds</li>
 * </ul>
 *
 * <p>Schedules can have conditions and multiple actions, supporting
 * complex automation scenarios.</p>
 *

 * @see User
 */
@Entity
@Table(name = "schedules", schema = "smartlighting")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Schedule {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(nullable = false, length = 100)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column
    @Builder.Default
    private Boolean enabled = true;

    /**
     * Type of trigger: 'time', 'sun', 'sensor'
     */
    @Column(name = "trigger_type", nullable = false, length = 20)
    private String triggerType;

    /**
     * Trigger configuration in JSON format.
     * For 'time': {"at": "07:00:00", "weekdays": ["mon","tue","wed","thu","fri"]}
     * For 'sun': {"event": "sunset", "offset_minutes": -30}
     * For 'sensor': {"entity": "luminosity", "above": 800}
     */
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "trigger_config", columnDefinition = "jsonb")
    @Builder.Default
    private Map<String, Object> triggerConfig = new HashMap<>();

    /**
     * Optional conditions that must be met for the schedule to execute.
     * Format: [{"type": "time", "after": "06:00", "before": "22:00"}]
     */
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "conditions", columnDefinition = "jsonb")
    @Builder.Default
    private List<Map<String, Object>> conditions = new ArrayList<>();

    /**
     * Actions to execute when triggered.
     * Format: [{"type": "scene", "scene_id": "..."}, {"type": "light", "target": "bedroom", "brightness": 50}]
     */
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "actions", columnDefinition = "jsonb")
    @Builder.Default
    private List<Map<String, Object>> actions = new ArrayList<>();

    @Column(name = "last_triggered_at")
    private LocalDateTime lastTriggeredAt;

    @Column(name = "trigger_count")
    @Builder.Default
    private Integer triggerCount = 0;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by")
    private User createdBy;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
}

File: mobile/app/src/main/java/com/smartlighting/mobile/data/model/Schedule.kt
package com.smartlighting.mobile.data.model

import com.google.gson.annotations.SerializedName

data class Schedule(
    @SerializedName("id")
    val id: String? = null,
    @SerializedName("name")
    val name: String,
    @SerializedName("description")
    val description: String? = null,
    @SerializedName("enabled")
    val enabled: Boolean = true,
    @SerializedName("triggerType")
    val triggerType: String,  // "time" or "sun"
    @SerializedName("triggerConfig")
    val triggerConfig: Map<String, Any>? = null,
    @SerializedName("actionType")
    val actionType: String? = null,  // "scene" or "device"
    @SerializedName("actionConfig")
    val actionConfig: Map<String, Any>? = null,
    @SerializedName("conditions")
    val conditions: List<Map<String, Any>>? = null,
    @SerializedName("actions")
    val actions: List<ScheduleAction>? = null,
    @SerializedName("lastTriggeredAt")
    val lastTriggeredAt: String? = null,
    @SerializedName("triggerCount")
    val triggerCount: Int? = 0,
    @SerializedName("createdAt")
    val createdAt: String? = null,
    @SerializedName("updatedAt")
    val updatedAt: String? = null,
    @SerializedName("createdBy")
    val createdBy: String? = null,
    @SerializedName("createdByName")
    val createdByName: String? = null
)

data class ScheduleAction(
    @SerializedName("type")
    val type: String? = null,
    @SerializedName("intent")
    val intent: String? = null,
    @SerializedName("target")
    val target: String? = null,
    @SerializedName("params")
    val params: Map<String, Any>? = null,
    @SerializedName("scene")
    val scene: String? = null,
    @SerializedName("scene_id")
    val sceneId: String? = null
)

data class NlpCommand(
    @SerializedName("text")
    val text: String? = null,
    @SerializedName("parsed")
    val parsed: ParsedCommand? = null,
    @SerializedName("valid")
    val valid: Boolean,
    @SerializedName("preview")
    val preview: String? = null,
    @SerializedName("error")
    val error: String? = null,
    @SerializedName("isScheduled")
    val isScheduled: Boolean? = false,
    @SerializedName("executed")
    val executed: Boolean? = false,
    @SerializedName("conflictAnalysis")
    val conflictAnalysis: ConflictAnalysis? = null,
    @SerializedName("result")
    val result: String? = null
)

data class ParsedCommand(
    @SerializedName("intent")
    val intent: String,
    @SerializedName("target")
    val target: String? = null,
    @SerializedName("params")
    val params: Map<String, Any>? = null,
    @SerializedName("scene")
    val scene: String? = null,
    @SerializedName("schedule")
    val schedule: ScheduleConfig? = null,
    @SerializedName("confidence")
    val confidence: Double? = null
)

/**
 * Schedule configuration parsed from natural language.
 * Matches backend's NlpCommandDto.ScheduleConfig
 */
data class ScheduleConfig(
    @SerializedName("time")
    val time: String? = null,           // Time in HH:MM format (for time triggers)
    @SerializedName("trigger")
    val trigger: String? = null,        // Sun event: "sunset" or "sunrise"
    @SerializedName("offsetMinutes")
    val offsetMinutes: Int? = null,     // Offset in minutes for sun events
    @SerializedName("recurrence")
    val recurrence: Any? = null         // Recurrence: "once", "daily", "weekdays", "weekends", or list of days
)

data class ConflictAnalysis(
    @SerializedName("hasConflicts")
    val hasConflicts: Boolean,
    @SerializedName("summary")
    val summary: String? = null,
    @SerializedName("conflicts")
    val conflicts: List<Conflict>? = null
)

data class Conflict(
    @SerializedName("scheduleId2")
    val scheduleId2: String,
    @SerializedName("scheduleName2")
    val scheduleName2: String,
    @SerializedName("severity")
    val severity: String,
    @SerializedName("description")
    val description: String,
    @SerializedName("resolutions")
    val resolutions: List<Resolution>? = null
)

data class Resolution(
    @SerializedName("id")
    val id: String,
    @SerializedName("description")
    val description: String,
    @SerializedName("changes")
    val changes: Map<String, Any>? = null
)

